mailer/mailme.go

```diff
@@ -0,0 +1,271 @@
+		bodyBytes, err := ioutil.ReadAll(resp.Body)
+		if err != nil && triesLeft > 0 {
+			return t.fetchTemplate(url, triesLeft-1)
+		}
+		if err != nil {
+			return "", err
+		}
+		return string(bodyBytes), err
+	}
+	if triesLeft > 0 {
+		return t.fetchTemplate(url, triesLeft-1)
+	}
+	return "", errors.New("Unable to fetch mail template")
+}
+
+func (m *MailmeMailer) MailBody(url string, defaultTemplate string, data map[string]interface{}) (string, error) {
+	if m.FuncMap == nil {
+		m.FuncMap = map[string]interface{}{}
+	}
+	if m.cache == nil {
+		m.cache = &TemplateCache{templates: map[string]*MailTemplate{}, funcMap: m.FuncMap}
+	}
+
+	var temp *template.Template
+	var err error
+
+	if url != "" {
+		var absoluteURL string
+		if strings.HasPrefix(url, "http") {
+			absoluteURL = url
+		} else {
+			absoluteURL = m.BaseURL + url
+		}
+		temp, err = m.cache.Get(absoluteURL)
+		if err != nil {
+			log.Printf("Error loading template from %v: %v\n", url, err)
+		}
+	}
+
+	if temp == nil {
+		cached, ok := m.cache.templates[url]
+		if ok {
+			temp = cached.tmp
+		} else {
+			temp, err = m.cache.Set(url, defaultTemplate, 0)
+			if err != nil {
+				return "", err
+			}
+		}
+	}
+
+	buf := &bytes.Buffer{}
+	err = temp.Execute(buf, data)
+	if err != nil {
+		return "", err
+	}
+	return buf.String(), nil
+}
+
+var privateIPBlocks []*net.IPNet
+
+func init() {
+	for _, cidr := range []string{
+		"127.0.0.0/8",    // IPv4 loopback
+		"10.0.0.0/8",     // RFC1918
+		"100.64.0.0/10",  // RFC6598
+		"172.16.0.0/12",  // RFC1918
+		"192.0.0.0/24",   // RFC6890
+		"192.168.0.0/16", // RFC1918
+		"169.254.0.0/16", // RFC3927
+		"::1/128",        // IPv6 loopback
+		"fe80::/10",      // IPv6 link-local
+		"fc00::/7",       // IPv6 unique local addr
+	} {
+		_, block, _ := net.ParseCIDR(cidr)
+		privateIPBlocks = append(privateIPBlocks, block)
+	}
+}
+
+func isPrivateIP(ip net.IP) bool {
+	for _, block := range privateIPBlocks {
+		if block.Contains(ip) {
+			return true
+		}
+	}
+	return false
+}
+
+type noLocalTransport struct {
+	inner  http.RoundTripper
+	errlog logrus.FieldLogger
+}
+
+func (no noLocalTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx, cancel := context.WithCancel(req.Context())
+
+	ctx = httptrace.WithClientTrace(ctx, &httptrace.ClientTrace{
+		ConnectStart: func(network, addr string) {
+			fmt.Printf("Checking network %v\n", addr)
+			host, _, err := net.SplitHostPort(addr)
+			if err != nil {
+				cancel()
+				fmt.Printf("Canceleing dur to error in addr parsing %v", err)
+				return
+			}
+			ip := net.ParseIP(host)
+			if ip == nil {
+				cancel()
+				fmt.Printf("Canceleing dur to error in ip parsing %v", host)
+				return
+			}
+
+			if isPrivateIP(ip) {
+				cancel()
+				fmt.Println("Canceleing dur to private ip range")
+				return
+			}
+
+		},
+	})
+
+	req = req.WithContext(ctx)
+	return no.inner.RoundTrip(req)
+}
+
+func SafeRountripper(trans http.RoundTripper, log logrus.FieldLogger) http.RoundTripper {
```

Comment:
There's a typo here

Reply:
Good catch! Updating (and all references)